features/steps/steps.py

```diff
@@ -221,3 +234,46 @@ def step_impl(context, related, relating, other_entity, condition):
                         errors.append(instance_structure_error(inst, rel.RelatingObject))
 
     handle_errors(context, errors)
+
+
+@then('Each {entity} {condition} be {directness} contained in {other_entity}')
+def step_impl(context, entity, condition, directness, other_entity):
+    stmt_to_op = ['must', 'must not']
+    assert condition in stmt_to_op
+
+    stmt_about_directness = ['directly', 'indirectly', 'directly or indirectly', 'indirectly or directly']
+    assert directness in stmt_about_directness
+
+    errors = []
+
+    if context.instances and getattr(context, 'applicable', True):
+        for ent in context.model.by_type(entity):
+            if len(ent.ContainedInStructure) > 0:
+                containing_relation = ent.ContainedInStructure[0]
+                relating_spatial_element = containing_relation.RelatingStructure
+                is_directly_contained = getattr(relating_spatial_element, 'is_a')(other_entity)
+                while len(relating_spatial_element.Decomposes) > 0:
+                    decomposed_element = relating_spatial_element.Decomposes[0]
+                    relating_spatial_element = decomposed_element.RelatingObject
+                    is_indirectly_contained = getattr(relating_spatial_element, 'is_a')(other_entity)
+                    if is_indirectly_contained:
+                        break
+            elif len(ent.ContainedInStructure) == 0:
+                is_directly_contained = False
+                is_indirectly_contained = False
+            if condition == 'must':
```

Comment:
I think these 6 cases can be generalized when thinking about them as set intersection.

You have:
- required_directness 
- observed_directness

if both of these are {'directly'}, {'indirectly'} or {'directly', 'indirectly'}. Then you can simply use the set intersection on both.

The set intersection should be empty in case of `must_not`, non empty in case of `must`.

`if bool(required_directness & observed_directness) == (condition == must):` <- untested but sth like this. It wouldn't win the price for readability so a comment is nice, or breaking this up into some variables can also help for readability. But it would greatly reduce the risk of errors when all cases are reduced to a single case.

Reply:
Very cool idea, added!